File: frontend/lib/api/timeline.ts
/**
 * Timeline APIs - Generate and fetch video timelines
 */

import { api } from "./client";
import { withLLMOverrides } from "./ai-overrides";
import type { TimelineResponse, TimelineGenerateResponse } from "./types";

export interface GetTimelineOptions {
    language?: string;
}

export interface GenerateTimelineOptions {
    /** Source language for loading subtitles */
    subtitleLanguage: string;
    /** Target language for LLM output */
    outputLanguage: string;
    learnerProfile?: string;
    force?: boolean;
}

/**
 * Fetch existing timeline (read-only).
 * Returns 404 if no cached timeline exists.
 */
export const getTimeline = async (
    contentId: string,
    options: GetTimelineOptions = {}
): Promise<TimelineResponse> => {
    const response = await api.get<TimelineResponse>(`/timeline/${contentId}`, {
        params: options.language ? { language: options.language } : undefined,
    });
    return response.data;
};

/**
 * Generate timeline using LLM.
 */
export const generateTimeline = async (
    contentId: string,
    options: GenerateTimelineOptions
): Promise<TimelineGenerateResponse> => {
    const response = await api.post<TimelineGenerateResponse>(
        `/timeline/${contentId}/generate`,
        withLLMOverrides({
            subtitle_language: options.subtitleLanguage,
            output_language: options.outputLanguage,
            learner_profile: options.learnerProfile,
            force: options.force ?? false,
        })
    );
    return response.data;
};

File: src/deeplecture/use_cases/dto/timeline.py
"""Timeline DTOs (Data Transfer Objects)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

# =============================================================================
# Request DTOs
# =============================================================================


@dataclass
class GenerateTimelineRequest:
    """Request to generate timeline.

    All language parameters must be provided by the frontend.
    """

    content_id: str
    subtitle_language: str  # Required: source language for loading subtitles
    output_language: str  # Required: target language for LLM output
    learner_profile: str | None = None
    force: bool = False
    # Runtime model/prompt selection (None = use defaults)
    llm_model: str | None = None
    prompts: dict[str, str] | None = None  # {func_id: impl_id}


# =============================================================================
# Response DTOs
# =============================================================================


@dataclass
class TimelineEntry:
    """
    A single timeline entry shown at a specific playback time.

    Represents a knowledge unit with explanation content.
    """

    id: int
    kind: str  # e.g., "segment_explanation"
    start: float  # Trigger time in seconds
    end: float  # End time in seconds
    title: str
    markdown: str

    def to_dict(self) -> dict[str, Any]:
        """Convert to JSON-serializable dict."""
        return {
            "id": self.id,
            "kind": self.kind,
            "start": self.start,
            "end": self.end,
            "title": self.title,
            "markdown": self.markdown,
        }


@dataclass
class TimelineResult:
    """Result of timeline generation."""

    content_id: str
    language: str
    entries: list[TimelineEntry]
    cached: bool = False
    status: str = "ready"

    @property
    def count(self) -> int:
        """Number of timeline entries."""
        return len(self.entries)

    def to_dict(self) -> dict[str, Any]:
        """Convert to JSON-serializable dict."""
        return {
            "video_id": self.content_id,  # Keep legacy API compatibility
            "language": self.language,
            "timeline": [entry.to_dict() for entry in self.entries],
            "count": self.count,
            "cached": self.cached,
            "status": self.status,
        }


# =============================================================================
# Internal DTOs (used within UseCase)
# =============================================================================


@dataclass
class SubtitleSegment:
    """
    Single subtitle entry with timing in seconds.

    Simple, self-contained representation for timeline generation.
    """

    id: int
    start: float
    end: float
    text: str

    @property
    def duration(self) -> float:
        """Segment duration in seconds."""
        return max(0.0, self.end - self.start)


@dataclass
class KnowledgeUnit:
    """
    A higher-level conceptual slice of the lecture.

    Represents a continuous time range covering one main idea.
    """

    id: int
    start: float
    end: float
    title: str
